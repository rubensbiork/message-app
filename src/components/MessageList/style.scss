.message-list-wrapper {
	.message-list {
		margin: 0;
	  padding: 0;
	  list-style: none;
	  min-height: 300px;
	  max-height: 300px;
	  overflow-y: auto;

		.no-message {
			font-size: 24px;
			width: calc(100% - 10px);
			border-radius: 10px;
			background: #eceff1;
			text-align: center;
			line-height: 4.4;
			color: #969696;
			font-weight: 400;
			box-shadow: 3px 3px 9px #968989;
		}
	}

	.counter {
		padding: 10px;
		display: inline-block;
		color: #969696;
		font-weight: 500;
		margin-bottom: 20px;
		border-radius: 5px;
		background: #eceff1;
		box-shadow: 3px 3px 9px #968989;
	}
}
