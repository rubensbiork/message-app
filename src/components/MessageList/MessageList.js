import React, { PureComponent } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';

import {
  createFragmentContainer,
  graphql,
} from 'react-relay';

import Message from '../Message';

import './style.scss';

export class MessageList extends PureComponent {

  componentDidUpdate(prevProps) {
    //it only scrolls to bottom when a new message is added
    if (this.props.viewer.totalCount > prevProps.viewer.totalCount) {
      this.scrollToBottom();
    }
  }

  scrollToBottom = () => {
    const node = ReactDOM.findDOMNode(this.messagesEnd);
    node.scrollIntoView({ behavior: 'smooth' });
  }

  noDataBlock = () => {
    return (
      <div className="no-message">
        No message here yet!
      </div>
    );
  }

  renderMessages() {
    const {
      edges,
    } = this.props.viewer.messages;

    if (edges.length === 0) {
      return this.noDataBlock();
    }

    return edges.map(edge =>
      <Message
        key={edge.node.id}
        message={edge.node}
        viewer={this.props.viewer}
      />
    );
  }

  renderCounter() {
    const {
      totalCount,
    } = this.props.viewer;

    return (
      <div className="counter">
        {totalCount} {parseInt(totalCount, 10) > 1 ? 'items' : 'item'}
      </div>
    );
  }

  render() {
    return (
      <div className="message-list-wrapper">
        {this.renderCounter()}
        <ul className="message-list">
          {this.renderMessages()}
          <div ref={(el) => { this.messagesEnd = el; }} />
        </ul>
      </div>
    );
  }
}

MessageList.defaultProps = {
  viewer: {},
};

MessageList.propTypes = {
  viewer: PropTypes.shape({}),
};

export default createFragmentContainer(MessageList, {
  viewer: graphql`
    fragment MessageList_viewer on User {
      messages(
        first: 100
      ) @connection(key: "MessageList_messages") {
        edges {
          node {
            id,
            ...Message_message,
          },
        },
      },
      id,
      totalCount,
      ...Message_viewer,
    }
  `,
});
