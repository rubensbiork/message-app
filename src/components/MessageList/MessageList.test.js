import React from 'react';
import { shallow } from 'enzyme';

import { MessageList } from './MessageList';

const wrapper = props => shallow(<MessageList {...props} />);

const messageData = {
  viewer: {
    totalCount: 2,
    messages: {
      edges: [
        {
          node: {
            id: 'some-id',
            text: 'some-text',
            createdAt: 'some-date',
          },
        },
        {
          node: {
            id: 'some-id2',
            text: 'some-text2',
            createdAt: 'some-date2',
          },
        },
      ],
    },
  },
};

describe('components/<MessageList />', () => {
  it('Should render', () => {
    const component = wrapper({...messageData});

    expect(
      component,
    ).toMatchSnapshot();
  });

  it('Should display the counter with 2 items', () => {
    const component = wrapper({...messageData});

    expect(
      component.find('.counter').text()
    ).toBe('2 items');
  });

  it('Should display no data message', () => {
    const noMessage = {
      viewer: {
        totalCount: 0,
        messages: {
          edges: [],
        },
      },
    };

    const component = wrapper({...noMessage});

    expect(
      component.find('.no-message').text()
    ).toBe('No message here yet!');

    expect(
      component.find('.no-message')
    ).toMatchSnapshot();
  });

});
