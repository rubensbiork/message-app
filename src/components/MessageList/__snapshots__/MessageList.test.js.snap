// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/<MessageList /> Should display no data message 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="no-message"
>
    No message here yet!
</div>,
  "nodes": Array [
    <div
      className="no-message"
>
      No message here yet!
</div>,
  ],
  "options": Object {},
  "renderer": null,
  "root": ShallowWrapper {
    "complexSelector": ComplexSelector {
      "buildPredicate": [Function],
      "childrenOfNode": [Function],
      "findWhereUnwrapped": [Function],
    },
    "length": 1,
    "node": <div
      className="message-list-wrapper"
>
      <div
            className="counter"
      >
            0
             
            item
      </div>
      <ul
            className="message-list"
      >
            <div
                  className="no-message"
            >
                  No message here yet!
            </div>
            <div />
      </ul>
</div>,
    "nodes": Array [
      <div
        className="message-list-wrapper"
>
        <div
                className="counter"
        >
                0
                 
                item
        </div>
        <ul
                className="message-list"
        >
                <div
                        className="no-message"
                >
                        No message here yet!
                </div>
                <div />
        </ul>
</div>,
    ],
    "options": Object {},
    "renderer": ReactShallowRenderer {
      "_instance": ShallowComponentWrapper {
        "_calledComponentWillUnmount": false,
        "_compositeType": 1,
        "_context": Object {},
        "_currentElement": <MessageList
          viewer={
                    Object {
                              "messages": Object {
                                "edges": Array [],
                              },
                              "totalCount": 0,
                            }
          }
/>,
        "_debugID": 5,
        "_hostContainerInfo": null,
        "_hostParent": null,
        "_instance": MessageList {
          "_reactInternalInstance": [Circular],
          "context": Object {},
          "noDataBlock": [Function],
          "props": Object {
            "viewer": Object {
              "messages": Object {
                "edges": Array [],
              },
              "totalCount": 0,
            },
          },
          "refs": Object {},
          "scrollToBottom": [Function],
          "state": null,
          "updater": Object {
            "enqueueCallback": [Function],
            "enqueueCallbackInternal": [Function],
            "enqueueElementInternal": [Function],
            "enqueueForceUpdate": [Function],
            "enqueueReplaceState": [Function],
            "enqueueSetState": [Function],
            "isMounted": [Function],
            "validateCallback": [Function],
          },
        },
        "_mountOrder": 3,
        "_pendingCallbacks": null,
        "_pendingElement": null,
        "_pendingForceUpdate": false,
        "_pendingReplaceState": false,
        "_pendingStateQueue": null,
        "_renderedComponent": NoopInternalComponent {
          "_currentElement": <div
            className="message-list-wrapper"
>
            <div
                        className="counter"
            >
                        0
                         
                        item
            </div>
            <ul
                        className="message-list"
            >
                        <div
                                    className="no-message"
                        >
                                    No message here yet!
                        </div>
                        <div />
            </ul>
</div>,
          "_debugID": 6,
          "_renderedOutput": <div
            className="message-list-wrapper"
>
            <div
                        className="counter"
            >
                        0
                         
                        item
            </div>
            <ul
                        className="message-list"
            >
                        <div
                                    className="no-message"
                        >
                                    No message here yet!
                        </div>
                        <div />
            </ul>
</div>,
        },
        "_renderedNodeType": 0,
        "_rootNodeID": 0,
        "_topLevelWrapper": null,
        "_updateBatchNumber": null,
        "_warnedAboutRefsInRender": false,
      },
      "getRenderOutput": [Function],
      "render": [Function],
    },
    "root": [Circular],
    "unrendered": <MessageList
      viewer={
            Object {
                  "messages": Object {
                    "edges": Array [],
                  },
                  "totalCount": 0,
                }
      }
/>,
  },
  "unrendered": null,
}
`;

exports[`components/<MessageList /> Should render 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="message-list-wrapper"
>
    <div
        className="counter"
    >
        2
         
        items
    </div>
    <ul
        className="message-list"
    >
        <Relay(Message)
            message={
                Object {
                    "createdAt": "some-date",
                    "id": "some-id",
                    "text": "some-text",
                  }
            }
            viewer={
                Object {
                    "messages": Object {
                      "edges": Array [
                        Object {
                          "node": Object {
                            "createdAt": "some-date",
                            "id": "some-id",
                            "text": "some-text",
                          },
                        },
                        Object {
                          "node": Object {
                            "createdAt": "some-date2",
                            "id": "some-id2",
                            "text": "some-text2",
                          },
                        },
                      ],
                    },
                    "totalCount": 2,
                  }
            }
        />
        <Relay(Message)
            message={
                Object {
                    "createdAt": "some-date2",
                    "id": "some-id2",
                    "text": "some-text2",
                  }
            }
            viewer={
                Object {
                    "messages": Object {
                      "edges": Array [
                        Object {
                          "node": Object {
                            "createdAt": "some-date",
                            "id": "some-id",
                            "text": "some-text",
                          },
                        },
                        Object {
                          "node": Object {
                            "createdAt": "some-date2",
                            "id": "some-id2",
                            "text": "some-text2",
                          },
                        },
                      ],
                    },
                    "totalCount": 2,
                  }
            }
        />
        <div />
    </ul>
</div>,
  "nodes": Array [
    <div
      className="message-list-wrapper"
>
      <div
            className="counter"
      >
            2
             
            items
      </div>
      <ul
            className="message-list"
      >
            <Relay(Message)
                  message={
                        Object {
                              "createdAt": "some-date",
                              "id": "some-id",
                              "text": "some-text",
                            }
                  }
                  viewer={
                        Object {
                              "messages": Object {
                                "edges": Array [
                                  Object {
                                    "node": Object {
                                      "createdAt": "some-date",
                                      "id": "some-id",
                                      "text": "some-text",
                                    },
                                  },
                                  Object {
                                    "node": Object {
                                      "createdAt": "some-date2",
                                      "id": "some-id2",
                                      "text": "some-text2",
                                    },
                                  },
                                ],
                              },
                              "totalCount": 2,
                            }
                  }
            />
            <Relay(Message)
                  message={
                        Object {
                              "createdAt": "some-date2",
                              "id": "some-id2",
                              "text": "some-text2",
                            }
                  }
                  viewer={
                        Object {
                              "messages": Object {
                                "edges": Array [
                                  Object {
                                    "node": Object {
                                      "createdAt": "some-date",
                                      "id": "some-id",
                                      "text": "some-text",
                                    },
                                  },
                                  Object {
                                    "node": Object {
                                      "createdAt": "some-date2",
                                      "id": "some-id2",
                                      "text": "some-text2",
                                    },
                                  },
                                ],
                              },
                              "totalCount": 2,
                            }
                  }
            />
            <div />
      </ul>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 1,
      "_context": Object {},
      "_currentElement": <MessageList
        viewer={
                Object {
                        "messages": Object {
                          "edges": Array [
                            Object {
                              "node": Object {
                                "createdAt": "some-date",
                                "id": "some-id",
                                "text": "some-text",
                              },
                            },
                            Object {
                              "node": Object {
                                "createdAt": "some-date2",
                                "id": "some-id2",
                                "text": "some-text2",
                              },
                            },
                          ],
                        },
                        "totalCount": 2,
                      }
        }
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": MessageList {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "noDataBlock": [Function],
        "props": Object {
          "viewer": Object {
            "messages": Object {
              "edges": Array [
                Object {
                  "node": Object {
                    "createdAt": "some-date",
                    "id": "some-id",
                    "text": "some-text",
                  },
                },
                Object {
                  "node": Object {
                    "createdAt": "some-date2",
                    "id": "some-id2",
                    "text": "some-text2",
                  },
                },
              ],
            },
            "totalCount": 2,
          },
        },
        "refs": Object {},
        "scrollToBottom": [Function],
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="message-list-wrapper"
>
          <div
                    className="counter"
          >
                    2
                     
                    items
          </div>
          <ul
                    className="message-list"
          >
                    <Relay(Message)
                              message={
                                        Object {
                                                  "createdAt": "some-date",
                                                  "id": "some-id",
                                                  "text": "some-text",
                                                }
                              }
                              viewer={
                                        Object {
                                                  "messages": Object {
                                                    "edges": Array [
                                                      Object {
                                                        "node": Object {
                                                          "createdAt": "some-date",
                                                          "id": "some-id",
                                                          "text": "some-text",
                                                        },
                                                      },
                                                      Object {
                                                        "node": Object {
                                                          "createdAt": "some-date2",
                                                          "id": "some-id2",
                                                          "text": "some-text2",
                                                        },
                                                      },
                                                    ],
                                                  },
                                                  "totalCount": 2,
                                                }
                              }
                    />
                    <Relay(Message)
                              message={
                                        Object {
                                                  "createdAt": "some-date2",
                                                  "id": "some-id2",
                                                  "text": "some-text2",
                                                }
                              }
                              viewer={
                                        Object {
                                                  "messages": Object {
                                                    "edges": Array [
                                                      Object {
                                                        "node": Object {
                                                          "createdAt": "some-date",
                                                          "id": "some-id",
                                                          "text": "some-text",
                                                        },
                                                      },
                                                      Object {
                                                        "node": Object {
                                                          "createdAt": "some-date2",
                                                          "id": "some-id2",
                                                          "text": "some-text2",
                                                        },
                                                      },
                                                    ],
                                                  },
                                                  "totalCount": 2,
                                                }
                              }
                    />
                    <div />
          </ul>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="message-list-wrapper"
>
          <div
                    className="counter"
          >
                    2
                     
                    items
          </div>
          <ul
                    className="message-list"
          >
                    <Relay(Message)
                              message={
                                        Object {
                                                  "createdAt": "some-date",
                                                  "id": "some-id",
                                                  "text": "some-text",
                                                }
                              }
                              viewer={
                                        Object {
                                                  "messages": Object {
                                                    "edges": Array [
                                                      Object {
                                                        "node": Object {
                                                          "createdAt": "some-date",
                                                          "id": "some-id",
                                                          "text": "some-text",
                                                        },
                                                      },
                                                      Object {
                                                        "node": Object {
                                                          "createdAt": "some-date2",
                                                          "id": "some-id2",
                                                          "text": "some-text2",
                                                        },
                                                      },
                                                    ],
                                                  },
                                                  "totalCount": 2,
                                                }
                              }
                    />
                    <Relay(Message)
                              message={
                                        Object {
                                                  "createdAt": "some-date2",
                                                  "id": "some-id2",
                                                  "text": "some-text2",
                                                }
                              }
                              viewer={
                                        Object {
                                                  "messages": Object {
                                                    "edges": Array [
                                                      Object {
                                                        "node": Object {
                                                          "createdAt": "some-date",
                                                          "id": "some-id",
                                                          "text": "some-text",
                                                        },
                                                      },
                                                      Object {
                                                        "node": Object {
                                                          "createdAt": "some-date2",
                                                          "id": "some-id2",
                                                          "text": "some-text2",
                                                        },
                                                      },
                                                    ],
                                                  },
                                                  "totalCount": 2,
                                                }
                              }
                    />
                    <div />
          </ul>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <MessageList
    viewer={
        Object {
            "messages": Object {
              "edges": Array [
                Object {
                  "node": Object {
                    "createdAt": "some-date",
                    "id": "some-id",
                    "text": "some-text",
                  },
                },
                Object {
                  "node": Object {
                    "createdAt": "some-date2",
                    "id": "some-id2",
                    "text": "some-text2",
                  },
                },
              ],
            },
            "totalCount": 2,
          }
    }
/>,
}
`;
